### PR TITLE
Add propagation to String strip methods

### PR DESCRIPTION
**Pull request by Mariovido**
# What Does This Do
This adds the instrumentation to propagate the taint values through the methods of `String` called `strip`, `stripLeading` and `stripTrailing`. This method are only available from version 11 of Java.

# Motivation
Increase propagation of String methods

# Additional Notes

# Contributor Checklist

- [x] Format the title [according the contribution guidelines](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#title-format)
- [x] Assign the `type:` and (`comp:` or `inst:`) labels in addition to [any usefull labels](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#labels)
- [x] Squash your commits prior merging or merge using GitHub's [Squash and merge](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squash-and-merge-your-commits)
- [x] Don't use `close`, `fix` or any [linking keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) when referencing an issue.  
  Use `solves` instead, and assign the PR [milestone](https://github.com/DataDog/dd-trace-java/milestones) to the issue
- [x] Update the [public documentation](https://docs.datadoghq.com/tracing/trace_collection/library_config/java/) in case of new configuration flag or behavior

Jira ticket: [APPSEC-5757]

[APPSEC-5757]: https://datadoghq.atlassian.net/browse/APPSEC-5757?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: dd-java-agent/agent-iast/src/test/groovy/com/datadog/iast/propagation/StringModuleTest.groovy [hunk 1]
```diff
@@ -984,6 +984,70 @@ class StringModuleTest extends IastModuleImplTestBase {
...
+    module.onStringStrip(self, result, trailing)
+    def taintedObject = taintedObjects.get(result)
+
+    then:
+    1 * tracer.activeSpan() >> span
+    taintFormat(result, taintedObject.getRanges()) == expected
+
+    where:
+    method          | trailing | testString        | expected
+    "strip"         | false    | "   ==>123<==   " | "==>123<=="
```
Comment (smola) [inline dd-java-agent/agent-iast/src/test/groovy/com/datadog/iast/propagation/StringModuleTest.groovy:1003]: Some things to test (in this or other test functions):
```
    "strip"         | false    | "   ==>123<==   " | "==>123<=="
    "strip"         | false    | "   ==>123<==   " | "==>123<=="
    "strip"         | false    | "==>   123   <==" | "==>123<=="
    "strip"         | false    | "   a==> b <==c   " | "a==> b <==c"
...
```

File: dd-java-agent/agent-iast/src/test/groovy/com/datadog/iast/propagation/StringModuleTest.groovy [hunk 2]
```diff
@@ -984,6 +984,70 @@ class StringModuleTest extends IastModuleImplTestBase {
...
+    "stripLeading"  | false    | "   ==>123<==   " | "==>123<==   "
+    "stripTrailing" | true     | "   ==>123<==   " | "   ==>123<=="
+  }
+
+  void 'test strip for not empty string cases'() {
```
Comment (smola) [inline dd-java-agent/agent-iast/src/test/groovy/com/datadog/iast/propagation/StringModuleTest.groovy:1008]: What's the difference with `test strip and make sure IastRequestContext is called`? Shouldn't this be just one test, with more cases in `where`?

File: dd-java-agent/instrumentation/java-lang/java-lang-11/src/test/groovy/datadog/trace/instrumentation/java/lang/jdk11/StringCallSiteTest.groovy [hunk 3]
```diff
@@ -31,4 +31,41 @@ class StringCallSiteTest extends AgentTestRunner {
...
+
+    then:
+    result == output
+    1 * module.onStringStrip(input, output)
+    0 * _
+
+    where:
+    method                | input     | output
+    "stringStrip"         | ' hello ' | 'hello'
+    "stringStripLeading"  | ' hello ' | 'hello '
```
Comment (smola) [inline dd-java-agent/instrumentation/java-lang/java-lang-11/src/test/groovy/datadog/trace/instrumentation/java/lang/jdk11/StringCallSiteTest.groovy:51]: stripTrailing not tested?

**Review by smola — CHANGES_REQUESTED** [review]

File: dd-java-agent/agent-iast/src/test/groovy/com/datadog/iast/propagation/StringModuleTest.groovy (hunk 1 above)
Comment (Mariovido) [reply dd-java-agent/agent-iast/src/test/groovy/com/datadog/iast/propagation/StringModuleTest.groovy]: Added those test

File: dd-java-agent/agent-iast/src/test/groovy/com/datadog/iast/propagation/StringModuleTest.groovy (hunk 2 above)
Comment (Mariovido) [reply dd-java-agent/agent-iast/src/test/groovy/com/datadog/iast/propagation/StringModuleTest.groovy]: There is not difference to be honest so I have merged them into one with multiple wheres

File: dd-java-agent/instrumentation/java-lang/java-lang-11/src/test/groovy/datadog/trace/instrumentation/java/lang/jdk11/StringCallSiteTest.groovy (hunk 3 above)
Comment (Mariovido) [reply dd-java-agent/instrumentation/java-lang/java-lang-11/src/test/groovy/datadog/trace/instrumentation/java/lang/jdk11/StringCallSiteTest.groovy:51]: I was creating an specific test for this one but It didn't make sense. I have merged them into one

**Review by smola — CHANGES_REQUESTED** [review]
Looks good. A couple of minor issues and we're good to go.

File: dd-java-agent/agent-iast/src/main/java/com/datadog/iast/propagation/StringModuleImpl.java [hunk 4]
```diff
@@ -567,6 +567,43 @@ public void onSplit(@Nonnull String self, @Nonnull String[] result) {
     }
   }
 
+  @Override
+  @SuppressFBWarnings
+  public void onStringStrip(@Nonnull String self, @Nonnull String result, boolean trailing) {
+    if (!canBeTainted(result) || self == result) {
```
[the remark this answers is not in the document]
Comment (Mariovido) [reply dd-java-agent/agent-iast/src/main/java/com/datadog/iast/propagation/StringModuleImpl.java]: Changed order

File: dd-java-agent/agent-iast/src/main/java/com/datadog/iast/propagation/StringModuleImpl.java [hunk 5]
```diff
@@ -567,6 +567,43 @@ public void onSplit(@Nonnull String self, @Nonnull String[] result) {
     }
   }
 
+  @Override
+  @SuppressFBWarnings
```
[the remark this answers is not in the document]
Comment (Mariovido) [reply dd-java-agent/agent-iast/src/main/java/com/datadog/iast/propagation/StringModuleImpl.java]: Added warning

**Review by smola — APPROVED** [review]

**Review by manuel-alvarez-alvarez — APPROVED** [review]
LGTM, nice one!

**Pull request merged**